src/main/scala/org/broadinstitute/dsde/firecloud/service/ShareLogService.scala

```diff
@@ -0,0 +1,57 @@
+package org.broadinstitute.dsde.firecloud.service
+
+import akka.pattern.pipe
+import akka.actor.{Actor, Props}
+import org.broadinstitute.dsde.firecloud.Application
+import org.broadinstitute.dsde.firecloud.model.ModelJsonProtocol.ShareFormat
+import org.broadinstitute.dsde.firecloud.dataaccess.ShareLogDAO
+import org.broadinstitute.dsde.firecloud.model.ShareLog.Share
+import org.broadinstitute.dsde.firecloud.model.UserInfo
+import org.broadinstitute.dsde.firecloud.service.PerRequest.{PerRequestMessage, RequestComplete}
+import org.broadinstitute.dsde.firecloud.service.ShareLogService._
+import org.broadinstitute.dsde.rawls.model.{AttributeFormat, PlainArrayAttributeListSerializer}
+import spray.httpx.SprayJsonSupport
+import spray.json.DefaultJsonProtocol._
+import scala.concurrent.{ExecutionContext, Future}
+
+object ShareLogService {
+  sealed trait ShareLogMessage
+  case class LogShare(userId: String, sharee: String, shareType: String) extends ShareLogMessage
+  case class GetShares(userId: String, shareType: Option[String] = None) extends ShareLogMessage
+  case class GetSharees(userId: String, shareType: Option[String] = None) extends ShareLogMessage
+//  case class Autocomplete(userId: String, term: String) extends ShareLogMessage
+
+  def props(constructor: () => ShareLogService): Props = {
+    Props(constructor())
+  }
+
+  def constructor(app: Application)(userInfo: UserInfo)(implicit executionContext: ExecutionContext) =
+    () => new ShareLogService(userInfo, app.shareLogDAO)
+}
+
+class ShareLogService(protected val userInfo: UserInfo, val shareDAO: ShareLogDAO)
+                     (implicit protected val executionContext: ExecutionContext) extends Actor with SprayJsonSupport {
+
+  implicit val impAttributeFormat: AttributeFormat = new AttributeFormat with PlainArrayAttributeListSerializer
+
+  override def receive = {
+    case LogShare(userId: String, sharee: String, shareType: String) => logShare(userId, sharee, shareType) pipeTo sender
+    case GetShares(userId: String, shareType: Option[String]) => getShares(userId) pipeTo sender
+    case GetSharees(userId: String, shareType: Option[String]) => getSharees(userId, shareType) pipeTo sender
+//    todo
+//    case Autocomplete(userId: String, term: String) => autocomplete(userId, term) pipeTo sender
+  }
+
+  def logShare(userId: String, sharee: String, shareType: String) = Future(RequestComplete(shareDAO.logShare(userId, sharee, shareType)))
```

Comment:
this will throw an error when attempting to log a duplicate entry. Is that desired?

As I type this, I wonder (scope creep warning) ... should we log the number of times a specific share (sharer, sharee) occurs? We could use that quantity to drive some kind of relevance - if Alice shared with Bob 20 times but only shared with Charlie twice, we could show Bob first.

Reply:
My initial thought was yes to throwing an error but now it doesn't seem practical to me, and I like your suggestion. I think ElasticSearch has functionality for this, I'll do some digging.